completion/all.go

```diff
@@ -0,0 +1,94 @@
+package completion
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/coder/serpent"
+)
+
+const (
+	BashShell  string = "bash"
+	FishShell  string = "fish"
+	ZShell     string = "zsh"
+	Powershell string = "powershell"
+)
+
+var shellCompletionByName = map[string]func(io.Writer, string) error{
+	BashShell:  GenerateCompletion(bashCompletionTemplate),
+	FishShell:  GenerateCompletion(fishCompletionTemplate),
+	ZShell:     GenerateCompletion(zshCompletionTemplate),
+	Powershell: GenerateCompletion(pshCompletionTemplate),
+}
+
+func ShellOptions(choice *string) *serpent.Enum {
+	return serpent.EnumOf(choice, BashShell, FishShell, ZShell, Powershell)
+}
+
+func GetCompletion(writer io.Writer, shell string, cmdName string) error {
+	fn, ok := shellCompletionByName[shell]
+	if !ok {
+		return fmt.Errorf("unknown shell %q", shell)
+	}
+	fn(writer, cmdName)
+	return nil
+}
+
+func GetUserShell() (string, error) {
```

Comment:
```suggestion
func UserShell() (string, error) {
```

A soft-verb such as `Detect` could be used here as well to help explain the API's probabilistic nature.